lib/hammer.ex

```diff
@@ -1,264 +1,60 @@
-  `{:ok, {count, count_remaining, ms_to_next_bucket, created_at, updated_at}`,
-  or `{:error, reason}`.
+      # Check the rate limit allowing 10 requests per second
+      MyApp.RateLimit.check_rate("some-key", _scale = :timer.seconds(1), _limit = 10)
 
-  Example:
+      # Wait for the next window
+      MyApp.RateLimit.wait(_scale = :timer.seconds(1))
 
-      inspect_bucket("file_upload:2042", 60_000, 5)
-      {:ok, {1, 2499, 29381612, 1450281014468, 1450281014468}}
+      # Reset the rate limiter
+      MyApp.RateLimit.reset("some-key", _scale = :timer.seconds(1))
 
   """
-  def inspect_bucket(id, scale_ms, limit) do
-    inspect_bucket(:single, id, scale_ms, limit)
-  end
-
-  @spec inspect_bucket(backend :: atom, id :: String.t(), scale_ms :: integer, limit :: integer) ::
-          {:ok,
-           {count :: integer, count_remaining :: integer, ms_to_next_bucket :: integer,
-            created_at :: integer | nil, updated_at :: integer | nil}}
-          | {:error, reason :: any}
-  @doc """
-  Same as inspect_bucket/3, but allows specifying a backend
-  """
-  def inspect_bucket(backend, id, scale_ms, limit) do
-    {stamp, key} = Utils.stamp_key(id, scale_ms)
-    ms_to_next_bucket = elem(key, 0) * scale_ms + scale_ms - stamp
-
-    case call_backend(backend, :get_bucket, [key]) do
-      {:ok, nil} ->
-        {:ok, {0, limit, ms_to_next_bucket, nil, nil}}
 
-      {:ok, {_, count, created_at, updated_at}} ->
-        count_remaining = if limit > count, do: limit - count, else: 0
-        {:ok, {count, count_remaining, ms_to_next_bucket, created_at, updated_at}}
-
-      {:error, reason} ->
-        {:error, reason}
-    end
-  end
+  @type key :: term
+  @type scale :: pos_integer
+  @type limit :: pos_integer
+  @type count :: pos_integer
+  @type increment :: non_neg_integer
 
-  @spec delete_buckets(id :: String.t()) ::
-          {:ok, count :: integer}
-          | {:error, reason :: any}
-  @doc """
-  Delete all buckets belonging to the provided id, including the current one.
-  Effectively resets the rate-limit for the id.
+  @callback hit(key, scale, increment) :: count
+  @callback set(key, scale, count) :: count
+  @callback get(key, scale) :: count
 
-  Arguments:
+  defmacro __using__(opts) do
+    quote bind_quoted: [opts: opts] do
+      @behaviour Hammer
+      @hammer_opts opts
 
-  - `id`: String name of the bucket
+      backend =
+        Keyword.get(opts, :backend) ||
+          raise ArgumentError, """
+          Hammer requires a backend to be specified. Example:
 
-  Returns either `{:ok, count}` where count is the number of buckets deleted,
-  or `{:error, reason}`.
+              use Hammer, backend: :ets
+          """
 
-  Example:
+      # this allows :ets to be aliased to Hammer.ETS
+      backend = with :ets <- backend, do: Hammer.ETS
+      @before_compile backend
 
-      user_id = 2406
-      {:ok, _count} = delete_buckets("file_uploads:\#{user_id}")
-
-  """
-  def delete_buckets(id) do
```

Comment:
why can t we have this anynore?

Reply:
Confusing name, bad performance. Instead, there is `MyApp.RateLimit.reset(key, scale)` now.